crates/bevy_math/src/ray_cast/dim2/polygon.rs

```diff
@@ -0,0 +1,157 @@
+use crate::prelude::*;
+
+// TODO: Polygons should probably have their own type for this along with a BVH acceleration structure.
+
+impl<const N: usize> PrimitiveRayCast2d for Polygon<N> {
+    #[inline]
+    fn local_ray_cast(&self, ray: Ray2d, max_distance: f32, solid: bool) -> Option<RayHit2d> {
+        local_ray_cast_polygon(&self.vertices, ray, max_distance, solid)
+    }
+}
+
+impl PrimitiveRayCast2d for BoxedPolygon {
+    #[inline]
+    fn local_ray_cast(&self, ray: Ray2d, max_distance: f32, solid: bool) -> Option<RayHit2d> {
+        local_ray_cast_polygon(&self.vertices, ray, max_distance, solid)
+    }
+}
+
+impl PrimitiveRayCast2d for RegularPolygon {
+    #[inline]
+    fn local_ray_cast(&self, ray: Ray2d, max_distance: f32, solid: bool) -> Option<RayHit2d> {
```

Comment:
I'm pretty sure we could avoid iterating all sides (at least for incoming rays, not for rays already near the polygon) by raycasting against the bounding (circumscribed) circle and then only checking the segment of the polygon corresponding to the point on the circle that got hit.

For raycasts already inside the bounding circle, then uh, I'm not sure, we can revert to the for loop behaviour.

Reply:
For rays inside the bounding circle, we could do a forwards and backwards raycast against the circle, and then check these two segments (which one is the real hit depends on whether the ray entered the polygon, or exited the polygon.)